server/lib/chat/chat-service.ts

```diff
@@ -280,15 +283,35 @@ export default class ChatService {
       }
     } while (!succeeded && !isUserBanned && attempts < MAX_JOIN_ATTEMPTS)
 
+    // NOTE(2Pac): This is just to silence the TS compiler since it can't figure out on its own that
+    // `channel` will be defined here.
+    channel = channel!
+
+    const joinChannelResponse: JoinChannelResponse = {
+      basicChannelInfo: {
+        id: channel.id,
+        name: channel.name,
+      },
+      detailedChannelInfo: {
+        private: channel.private,
+        official: channel.official,
+        userCount: channel.userCount,
+      },
+      joinedChannelInfo: {
+        ownerId: channel.ownerId,
+        topic: channel.topic,
+      },
+    }
+
     if (isUserInChannel) {
-      return channel!
+      return joinChannelResponse
     } else if (isUserBanned) {
       throw new ChatServiceError(ChatServiceErrorCode.UserBanned, 'User is banned')
```

Comment:
imo this check/throw should be above constructing `joinChannelResponse` now, probably just check it right after the loop

Reply:
Could probably restructure the code following this as well to avoid needing to construct this object before the return entirely:

```ts
if (!isUserInChannel) {
  this.updateUserAfterJoining(/* ... */)
}

return {
  // ...
}
```